### ISSUE TITLE
Support for sealed hierarchy of case classes

### ISSUE THREAD

**Issue by edmondo1984** [sksamuel/avro4s#176]
Using ADT in Scala is a common pattern and they are typically implemented through sealed hierarchy of case classes. At the moment it is not possible to auto-generate the necessary typeclasses in Avro4s, although some of the necessary infrastructure is already available through the usage of shapeless Coproduct.

## Current macro design
We got a little confused about the design of the macros, in particularly inside SchemaFormat and ToRecord:

- There is one implicit materializer and a method named `apply`, which always invokes the macro
- Macros using other macros do not rely on already available implicits but they re-invoke the macro themselves.

Therefore, the following problem can occurs:
- An explicitly defined typeclasses of type SchemaFor[A] is ignored by ToRecord[A] which invokes SchemaFor.apply, always going through the macro expansion
- There is a problem with macro priorities, because the macro materialization is the one with highest priority, while it should be the one with the lowest priority.

## Improvements in syntax / fix to priorities

We therefore suggest a design like the following:

1. The method apply should not perform any macro invocation but it should just summon the implicit, a-la-shapeless / cats or like many other libraries do
2. The implicit macro materializer should go into a trait so to have lower priority both with respect to

     1. Explicitly defined instances of the typeclasses 
     2. Combinators for collections or other types (sealed hierarchy of case classes, using recursive typeclass generation)

```scala

trait LowerPriorityImplicits {
      implicit def materializeSchemaFor[T]:SchemaFor[T] = macro SchemaFor.applyImpl[T]
}
object SchemaFor extends LowerPriorityImplicits{

  implicit object ByteSchemaFor extends SchemaFor[Byte] {
    private val schema = SchemaBuilder.builder().intType()
    def apply(): org.apache.avro.Schema = schema
  }

  def apply[T](schemaFor:SchemaFor[T]): SchemaFor[T] = schemaFor

  def applyImpl[T: c.WeakTypeTag](c: whitebox.Context): c.Expr[SchemaFor[T]] 

}

```

## Reusing already defined implicits

The ToRecord macro require an instance of SchemaFor, but instead of retrieving it from the context it explicitly invokes the macro ignoring implicits already in scope:

```scala

c.Expr[ToRecord[T]](
        q"""new _root_.com.sksamuel.avro4s.ToRecord[$tpe] {
            private val schemaFor : _root_.com.sksamuel.avro4s.SchemaFor[$tpe] = _root_.com.sksamuel.avro4s.SchemaFor[$tpe]
            private val converters : Array[_root_.shapeless.Lazy[_root_.com.sksamuel.avro4s.ToValue[_]]] = Array(..$converters)

            def apply(t : $tpe): _root_.org.apache.avro.generic.GenericRecord = {

              val record = new _root_.org.apache.avro.generic.GenericData.Record(schemaFor())
              ..$puts
              record
            }
          }"""

```

Instead, since _apply_ and the _materializer_  are defined on the companion object of SchemaFor, the implicit should be inferred from the context: if a concrete instance is available it will be picked up, if not it will fall back to the macro

```scala
  val schemaForImplicit = c.inferImplicitValue(tct, silent = true, withMacrosDisabled = true)
    if(schemaForImplicit == EmptyTree){
      c.abort(c.enclosingPosition,s"Cannot find Implicit SchemaFor for type $tct")
    }
    c.Expr[ToRecord[T]](
      q"""new _root_.com.sksamuel.avro4s.ToRecord[$tpe] {
            private val schemaFor : _root_.com.sksamuel.avro4s.SchemaFor[$tpe] = $schemaForImplicit
            private val converters : Array[_root_.shapeless.Lazy[_root_.com.sksamuel.avro4s.ToValue[_]]] = Array(..$converters)

            def apply(t : $tpe): _root_.org.apache.avro.generic.GenericRecord = {

              val record = new _root_.org.apache.avro.generic.GenericData.Record(schemaFor())
              ..$puts
              record
            }
          }"""
    )
```scala

**Comment by edmondo1984** [comment]
We have a branch  where we are working to fix the issue https://github.com/credimi/avro4s/tree/wip

With the new design invoking SchemaFor[SealedTrait] works correctly but the tests are not compiling because other implicits cannot be found.

[error] /Users/edmondoporcu1/Development/avro4s/avro4s-core/src/test/scala/com/sksamuel/avro4s/ReadmeTest.scala:34: Could not find implicit SchemaFor[Seq[com.sksamuel.avro4s.Ingredient]]
[error]       implicit val val2 = SchemaFor[Seq[Ingredient]]

We have noticed that inside SchemaFor, there are no collection combinators: why so? How are SchemaFor and ToSchema resolution related?

**Comment by sksamuel** [comment]
Your comments about using implicits over explicit macro expansion make sense, and I'm happy for you (or me) to correct that.

I don't understand your question re: "collection combinators" though, can you explain in more detail? 

SchemaFor and ToSchema are the equivalent typeclasses for avro's Schema, Schema.Field and Schema.Type basically.

**Comment by edmondo1984** [comment]
Collection combinators are like so:

`implicit def schemaForList[T](implicit schemaFor:SchemaFor[T]):SchemaFor[List[T]]`

what we do not understand is why there are such for ToSchema but not in SchemaFor

**Comment by sksamuel** [comment]
I don't check the gitter channel anymore, it's a relic.

The reason there's no schema for lists, is because schema for generates schemas for records. A record schema has fields, and each field gets its schema from ToSchema.

**Comment by edmondo1984** [comment]
@sksamuel I am blocked by https://github.com/scala/bug/issues/10745 if you have any useful hint would really appreciate

**Comment by edmondo1984** [comment]
We actually now see there is support for sealed hierarchies when it comes to Values and not to Records, which is already a really good thing, thank you. We might be able to workaround the issue we have in this case, however we are stucked by a sealed trait with mixed case objects and case classes as children. Is the following supposed to compile?

  ```val abc1d = FromValue.CoproductFromValue[Accept.type, CNil]```

**Comment by edmondo1984** [comment]
We investigated the issue further and compared how to Avro4s macros are written and we compared them with circe ones. This is how circe works:

- Macro generates conversions for HLists and Coproducts
- Implicit conversions creates json serializer / deserializer for any T where there is a Generic[T] with Repr = Hlist or Coproduct for which a Macro can generate a proper serializer/deserializer

These macros can handle cases such as the one I have described above very simply. For now I would put this issue on hold as with relative little boilerplate we were able to get it to work. As an overall improvement I think it would make sense to make the macro looks like circe ones (we have done some work we are happy to share)
